File: src/mac/table.m
// Rows and columns, filled by asking rather than by building.
//
// The one control in this host that calls *into* Beans. The reasoning is in
// ../cortado_host.h beside ctd_table_fn; what is here is AppKit's half of it:
// an NSTableView inside an NSScrollView, and one object standing in as its
// data source and delegate.

#import "internal.h"
#import <objc/runtime.h>

// One source for the process, like the event sink, and routed the same way:
// on the table's handle. Not one per table — a stored callback per control is
// a leak by construction, which the hazard note at the top of the header
// states in full.
static ctd_table_fn g_table_source;
static void        *g_table_context;
static ctd_table_editable_fn g_table_edit_policy;
static void                  *g_table_edit_context;

// The text of one cell, asked for the way ctd_get_text answers: write at most
// `cap`, answer what was needed. A 256-byte stack buffer covers a cell in
// almost every table there has ever been, and the heap path is there for the
// one that does not — truncating instead would be a table that is subtly wrong
// exactly where the data is interesting.
static NSString *ctd_table_text(ctd_handle table, int32_t row, int32_t column) {
    if (!g_table_source) return @"";
    char small[256];
    int32_t needed = g_table_source(g_table_context, table, row, column,
                                    small, (int32_t)sizeof small);
    if (needed <= 0) return @"";
    if (needed <= (int32_t)sizeof small) {
        NSString *text = [[[NSString alloc] initWithBytes:small
                                                  length:(NSUInteger)needed
                                                encoding:NSUTF8StringEncoding] autorelease];
        return text ? text : @"";
    }
    char *big = (char *)malloc((size_t)needed);
    if (!big) return @"";
    int32_t wrote = g_table_source(g_table_context, table, row, column, big, needed);
    NSString *text = nil;
    if (wrote > 0) {
        text = [[[NSString alloc] initWithBytes:big
                                         length:(NSUInteger)(wrote < needed ? wrote : needed)
                                       encoding:NSUTF8StringEncoding] autorelease];
    }
    free(big);
    return text ? text : @"";
}

@implementation CortadoDataTable
- (BOOL)ctdCompact {
    return [objc_getAssociatedObject(self, @selector(ctdCompact)) boolValue];
}
- (void)drawBackgroundInClipRect:(NSRect)clip {
    if (![self ctdCompact]) { [super drawBackgroundInClipRect:clip]; return; }
    BOOL dark = [[[self effectiveAppearance] bestMatchFromAppearancesWithNames:@[NSAppearanceNameAqua, NSAppearanceNameDarkAqua]] isEqualToString:NSAppearanceNameDarkAqua];
    [[NSColor colorWithCalibratedWhite:dark ? 0.12 : 1.0 alpha:1] setFill];
    NSRectFill(clip);
    CGFloat pitch = [self rowHeight] + [self intercellSpacing].height;
    NSInteger first = MAX(0, (NSInteger)floor(NSMinY(clip) / pitch));
    NSInteger last = (NSInteger)ceil(NSMaxY(clip) / pitch);
    [[NSColor colorWithCalibratedWhite:dark ? 0.155 : 0.96 alpha:1] setFill];
    for (NSInteger row = first; row <= last; row++) {
        if (row % 2) NSRectFill(NSIntersectionRect(clip, NSMakeRect(0, row * pitch, [self bounds].size.width, pitch)));
    }
}
- (void)mouseDown:(NSEvent *)event {
    NSPoint point = [self convertPoint:[event locationInWindow] fromView:nil];
    NSInteger column = [self columnAtPoint:point];
    if (column >= 0) _ctdActiveColumn = column;
    [super mouseDown:event];
    [self setNeedsDisplay:YES];
}
- (void)drawRect:(NSRect)dirty {
    [super drawRect:dirty];
    if (![self ctdCompact] || [self selectedRow] < 0 ||
        _ctdActiveColumn < 0 || _ctdActiveColumn >= [self numberOfColumns]) return;
    NSRect cell = [self frameOfCellAtColumn:_ctdActiveColumn row:[self selectedRow]];
    [[NSColor keyboardFocusIndicatorColor] setStroke];
    NSBezierPath *ring = [NSBezierPath bezierPathWithRect:NSInsetRect(cell, 1, 1)];
    [ring setLineWidth:2];
    [ring stroke];
}
- (void)copy:(id)sender {
    if (![self ctdCompact] || [self selectedRow] < 0 || [self numberOfColumns] == 0) return;
    NSInteger col = MIN(MAX(0, _ctdActiveColumn), [self numberOfColumns] - 1);
    id value = [[self dataSource] tableView:self objectValueForTableColumn:[[self tableColumns] objectAtIndex:col] row:[self selectedRow]];
    if (![value isKindOfClass:[NSString class]]) return;
    NSPasteboard *pasteboard = [NSPasteboard generalPasteboard];
    [pasteboard clearContents];
    [pasteboard setString:value forType:NSPasteboardTypeString];
}
- (void)keyDown:(NSEvent *)event {
    if (![self ctdCompact]) { [super keyDown:event]; return; }
    NSString *key = [event charactersIgnoringModifiers];
    if (([event modifierFlags] & NSEventModifierFlagCommand) && [key isEqualToString:@"c"]) {
        [self copy:nil]; return;
    }
    NSInteger row = [self selectedRow];
    NSInteger cols = [self numberOfColumns];
    if (row < 0 || cols == 0) { [super keyDown:event]; return; }
    _ctdActiveColumn = MIN(MAX(0, _ctdActiveColumn), cols - 1);
    unsigned short code = [event keyCode];
    if (code == 123 || code == 124 || code == 48) {
        NSInteger step = code == 123 || (code == 48 && ([event modifierFlags] & NSEventModifierFlagShift)) ? -1 : 1;
        NSInteger col = _ctdActiveColumn + step;
        if (code == 48 && col >= cols && row + 1 < [self numberOfRows]) { row++; col = 0; }
        if (code == 48 && col < 0 && row > 0) { row--; col = cols - 1; }
        _ctdActiveColumn = MIN(MAX(0, col), cols - 1);
        [self selectRowIndexes:[NSIndexSet indexSetWithIndex:row] byExtendingSelection:NO];
        [self scrollRowToVisible:row];
        [self scrollColumnToVisible:_ctdActiveColumn];
        [self setNeedsDisplay:YES];
        return;
    }
    if (code == 36 || code == 76) {
        NSTableColumn *column = [[self tableColumns] objectAtIndex:_ctdActiveColumn];
        id delegate = [self delegate];
        if ([delegate respondsToSelector:@selector(tableView:shouldEditTableColumn:row:)] &&
            [delegate tableView:self shouldEditTableColumn:column row:row]) {
            [self editColumn:_ctdActiveColumn row:row withEvent:nil select:YES];
        }
        return;
    }
    [super keyDown:event];
    [self setNeedsDisplay:YES];
}
@end

@implementation CortadoTableSource
- (NSInteger)numberOfRowsInTableView:(NSTableView *)view {
    (void)view;
    return _rows;
}

- (id)tableView:(NSTableView *)view
    objectValueForTableColumn:(NSTableColumn *)column
                          row:(NSInteger)row {
    NSInteger at = [[view tableColumns] indexOfObject:column];
    if (at == NSNotFound) return @"";
    return ctd_table_text(_handle, (int32_t)row, (int32_t)at);
}

- (BOOL)tableView:(NSTableView *)view
    shouldEditTableColumn:(NSTableColumn *)column
                     row:(NSInteger)row {
    NSInteger at = [[view tableColumns] indexOfObject:column];
    if (!_editable || !g_table_edit_policy || at == NSNotFound ||
        row < 0 || row >= _rows) return NO;
    return g_table_edit_policy(g_table_edit_context, _handle,
                               (int32_t)row, (int32_t)at) != 0;
}

- (void)tableView:(NSTableView *)view
    setObjectValue:(id)value
    forTableColumn:(NSTableColumn *)column
              row:(NSInteger)row {
    NSInteger at = [[view tableColumns] indexOfObject:column];
    if (at == NSNotFound || row < 0 || row >= _rows) return;
    if (![self tableView:view shouldEditTableColumn:column row:row]) {
        [view reloadDataForRowIndexes:[NSIndexSet indexSetWithIndex:(NSUInteger)row]
                       columnIndexes:[NSIndexSet indexSetWithIndex:(NSUInteger)at]];
        return;
    }
    NSString *text = [value isKindOfClass:[NSString class]] ? value : @"";
    const char *bytes = [text UTF8String];
    NSUInteger byteCount = [text lengthOfBytesUsingEncoding:NSUTF8StringEncoding];
    if (!bytes || byteCount > INT32_MAX || ctd_has_nul(bytes, (int32_t)byteCount)) return;
    int32_t length = (int32_t)byteCount;
    ctd_handle handle = _handle;
    [view retain];
    [text retain];
    if (g_sink && !g_writing) {
        ctd_event event;
        memset(&event, 0, sizeof event);
        event.kind = CTD_EV_TEXT_COMMIT;
        event.target = handle;
        event.index = (int64_t)row;
        event.token = (int64_t)at;
        event.text = bytes;
        event.text_len = length;
        g_sink(g_sink_context, &event);
    }
    // The handler may close the table or replace its source. Read back only
    // while the handle remains live; accepted and rejected edits both redraw
    // from the source instead of keeping an unsaved editor value.
    if (ctd_resolve(handle) && row < [view numberOfRows] &&
        at < (NSInteger)[[view tableColumns] count]) {
        [view reloadDataForRowIndexes:[NSIndexSet indexSetWithIndex:(NSUInteger)row]
                       columnIndexes:[NSIndexSet indexSetWithIndex:(NSUInteger)at]];
    }
    [text release];
    [view release];
}

- (void)tableViewSelectionDidChange:(NSNotification *)note {
    // Not when the program did it. NSTableView posts this for
    // -selectRowIndexes: exactly as it does for a click, and the header's
    // rule is that a write is silent — see g_writing in internal.h. A
    // navigator that opens a node and then selects it would otherwise hear
    // the selection and toggle the node shut again, which is what
    // examples/cask did.
    if (g_writing) return;
    NSTableView *view = (NSTableView *)[note object];
    NSInteger row = [view isKindOfClass:[NSTableView class]] ? [view selectedRow] : -1;
    ctd_emit(CTD_EV_SELECTION, _handle, (int64_t)row, 0);
}
@end

// The NSTableView a handle stands for. cortado tracks the *scroll view*,
// because that is the thing with a frame — a table that cannot scroll is a
// table with a hidden bottom — and every call here reaches through it.
NSTableView *ctd_table_view(id object) {
    if (![object isKindOfClass:[NSScrollView class]]) return nil;
    id inner = [(NSScrollView *)object documentView];
    if ([inner isKindOfClass:[NSTableView class]]) return (NSTableView *)inner;
    return nil;
}

// -------------------------------------------------------------- entry points

ctd_status ctd_set_table_source(ctd_table_fn source, void *context) {
    g_table_source = source;
    g_table_context = context;
    return CTD_OK;
}

ctd_status ctd_set_table_edit_policy(ctd_table_editable_fn policy, void *context) {
    g_table_edit_policy = policy;
    g_table_edit_context = context;
    return CTD_OK;
}

ctd_status ctd_table_columns(ctd_handle table, int32_t count) {
    id object = ctd_resolve(table);
    if (!object) return CTD_ERR_STALE;
    NSTableView *view = ctd_table_view(object);
    if (!view) return CTD_ERR_KIND;
    if (count < 0) return CTD_ERR_RANGE;
    while ([[view tableColumns] count] > 0) {
        [view removeTableColumn:[[view tableColumns] objectAtIndex:0]];
    }
    for (int32_t i = 0; i < count; i++) {
        NSString *name = [NSString stringWithFormat:@"c%d", i];
        NSTableColumn *column = [[NSTableColumn alloc] initWithIdentifier:name];
        [[column headerCell] setStringValue:@""];
        CortadoTableSource *source = (CortadoTableSource *)[view dataSource];
        BOOL editing = source && [source editable];
        [column setEditable:editing];
        if ([objc_getAssociatedObject(view, @selector(ctdCompact)) boolValue]) {
            [[column dataCell] setFont:[NSFont monospacedSystemFontOfSize:12.0 weight:NSFontWeightRegular]];
            [[column dataCell] setLineBreakMode:NSLineBreakByTruncatingTail];
        }
        [[column dataCell] setSelectable:YES];
        [[column dataCell] setEditable:editing];
        [view addTableColumn:column];
        [column release];
    }
    [view reloadData];
    return CTD_OK;
}

ctd_status ctd_table_editing(ctd_handle table, int32_t on) {
    id object = ctd_resolve(table);
    if (!object) return CTD_ERR_STALE;
    NSTableView *view = ctd_table_view(object);
    if (!view) return CTD_ERR_KIND;
    if (on != 0 && on != 1) return CTD_ERR_RANGE;
    CortadoTableSource *source = (CortadoTableSource *)[view dataSource];
    if (!source) return CTD_ERR_STATE;
    [source setEditable:on != 0];
    for (NSTableColumn *column in [view tableColumns]) {
        [column setEditable:on != 0];
        [[column dataCell] setSelectable:YES];
        [[column dataCell] setEditable:on != 0];
    }
    return CTD_OK;
}

ctd_status ctd_table_edit_as_user(ctd_handle table, int32_t row, int32_t column,
                                  const char *utf8, int32_t len) {
    if (len < 0 || (!utf8 && len > 0) || ctd_has_nul(utf8, len))
        return CTD_ERR_RANGE;
    id object = ctd_resolve(table);
    if (!object) return CTD_ERR_STALE;
    NSTableView *view = ctd_table_view(object);
    if (!view) return CTD_ERR_KIND;
    if (row < 0 || row >= [view numberOfRows] ||
        column < 0 || column >= (int32_t)[[view tableColumns] count]) return CTD_ERR_RANGE;
    CortadoTableSource *source = (CortadoTableSource *)[view dataSource];
    NSTableColumn *at = [[view tableColumns] objectAtIndex:(NSUInteger)column];
    if (!source || ![source tableView:view shouldEditTableColumn:at row:row])
        return CTD_ERR_STATE;
    [source tableView:view setObjectValue:ctd_string(utf8, len)
                           forTableColumn:at row:row];
    return CTD_OK;
}

ctd_status ctd_table_column_title(ctd_handle table, int32_t column,
                                  const char *utf8, int32_t len) {
    if (ctd_has_nul(utf8, len)) return CTD_ERR_RANGE;
    id object = ctd_resolve(table);
    if (!object) return CTD_ERR_STALE;
    NSTableView *view = ctd_table_view(object);
    if (!view) return CTD_ERR_KIND;
    if (column < 0 || column >= (int32_t)[[view tableColumns] count]) return CTD_ERR_RANGE;
    [[[[view tableColumns] objectAtIndex:column] headerCell]
        setStringValue:ctd_string(utf8, len)];
    return CTD_OK;
}

ctd_status ctd_table_column_width(ctd_handle table, int32_t column, double points) {
    id object = ctd_resolve(table);
    if (!object) return CTD_ERR_STALE;
    NSTableView *view = ctd_table_view(object);
    if (!view) return CTD_ERR_KIND;
    if (column < 0 || column >= (int32_t)[[view tableColumns] count]) return CTD_ERR_RANGE;
    if (points <= 0.0) return CTD_ERR_RANGE;
    [[[view tableColumns] objectAtIndex:column] setWidth:(CGFloat)points];
    return CTD_OK;
}

ctd_status ctd_table_rows(ctd_handle table, int32_t count) {
    id object = ctd_resolve(table);
    if (!object) return CTD_ERR_STALE;
    NSTableView *view = ctd_table_view(object);
    if (!view) return CTD_ERR_KIND;
    if (count < 0) return CTD_ERR_RANGE;
    CortadoTableSource *source = (CortadoTableSource *)[view dataSource];
    if (!source) return CTD_ERR_STATE;
    [source setRows:count];
    [view reloadData];
    return CTD_OK;
}

ctd_status ctd_table_reload(ctd_handle table) {
    id object = ctd_resolve(table);
    if (!object) return CTD_ERR_STALE;
    NSTableView *view = ctd_table_view(object);
    if (!view) return CTD_ERR_KIND;
    [view reloadData];
    return CTD_OK;
}

int32_t ctd_table_cell(ctd_handle table, int32_t row, int32_t column,
                       char *out, int32_t cap) {
    id object = ctd_resolve(table);
    if (!object) return CTD_ERR_STALE;
    NSTableView *view = ctd_table_view(object);
    if (!view) return CTD_ERR_KIND;
    if (row < 0 || row >= [view numberOfRows]) return CTD_ERR_RANGE;
    NSArray *columns = [view tableColumns];
    if (column < 0 || column >= (int32_t)[columns count]) return CTD_ERR_RANGE;
    // Through the data source, not around it: asking cortado's own book again
    // would prove only that cortado agrees with itself.
    id value = [[view dataSource] tableView:view
                  objectValueForTableColumn:[columns objectAtIndex:column]
                                        row:row];
    NSString *text = [value isKindOfClass:[NSString class]] ? (NSString *)value : @"";
    return ctd_copy_out(text, out, cap);
}

ctd_status ctd_table_selected(ctd_handle table, int32_t *out) {
    id object = ctd_resolve(table);
    if (!object) return CTD_ERR_STALE;
    NSTableView *view = ctd_table_view(object);
    if (!view) return CTD_ERR_KIND;
    if (out) *out = (int32_t)[view selectedRow];
    return CTD_OK;
}

ctd_status ctd_table_select(ctd_handle table, int32_t row) {
    id object = ctd_resolve(table);
    if (!object) return CTD_ERR_STALE;
    NSTableView *view = ctd_table_view(object);
    if (!view) return CTD_ERR_KIND;
    if (row < 0) {
        g_writing = g_writing + 1;
        [view deselectAll:nil];
        g_writing = g_writing - 1;
        return CTD_OK;
    }
    if (row >= [view numberOfRows]) return CTD_ERR_RANGE;
    g_writing = g_writing + 1;
    [view selectRowIndexes:[NSIndexSet indexSetWithIndex:(NSUInteger)row]
      byExtendingSelection:NO];
    g_writing = g_writing - 1;
    return CTD_OK;
}

File: src/win32/table.c
// Rows and columns, filled by asking rather than by building.
//
// Win32's half of the contract in ../cortado_host.h, and the platform whose
// own name for this is the clearest: a list view with LVS_OWNERDATA holds no
// items at all. It knows how many rows there are and sends the parent an
// LVN_GETDISPINFO for the text of a cell it is about to paint. A million rows
// costs a million nothing.

#include "internal.h"

static ctd_table_fn g_table_source;
static void        *g_table_context;

// One cell, as UTF-16 for the control. The caller owns the returned buffer and
// frees it; ctd_wide is the same arrangement everywhere else in this host.
static WCHAR *ctd_table_wide(ctd_handle table, int32_t row, int32_t column) {
    if (!g_table_source) return NULL;
    char small[256];
    int32_t needed = g_table_source(g_table_context, table, row, column,
                                    small, (int32_t)sizeof small);
    if (needed <= 0) return NULL;
    if (needed <= (int32_t)sizeof small) return ctd_wide(small, needed);
    char *big = (char *)malloc((size_t)needed);
    if (!big) return NULL;
    int32_t wrote = g_table_source(g_table_context, table, row, column, big, needed);
    WCHAR *wide = wrote > 0 ? ctd_wide(big, wrote < needed ? wrote : needed) : NULL;
    free(big);
    return wide;
}

// What the control asks for when it is about to paint a cell. Called from the
// parent's WM_NOTIFY, which is where every Win32 control reports.
void ctd_table_disp_info(NMLVDISPINFOW *info) {
    if (!info || !(info->item.mask & LVIF_TEXT)) return;
    if (!info->item.pszText || info->item.cchTextMax <= 0) return;
    ctd_handle table = ctd_handle_of(info->hdr.hwndFrom);
    if (!table) return;
    info->item.pszText[0] = L'\0';
    WCHAR *text = ctd_table_wide(table, (int32_t)info->item.iItem,
                                 (int32_t)info->item.iSubItem);
    if (!text) return;
    // The control owns the buffer and says how big it is; anything longer is
    // cut here rather than written past the end.
    lstrcpynW(info->item.pszText, text, info->item.cchTextMax);
    free(text);
}

// A selection moved. LVN_ITEMCHANGED fires for every state change, so the
// ones that are not a selection arriving are dropped.
void ctd_table_item_changed(NMLISTVIEW *info) {
    if (!info) return;
    // Not when the program did it — see g_writing in internal.h.
    if (g_writing) return;
    if (!(info->uChanged & LVIF_STATE)) return;
    if (!(info->uNewState & LVIS_SELECTED)) return;
    if (info->uOldState & LVIS_SELECTED) return;
    ctd_handle table = ctd_handle_of(info->hdr.hwndFrom);
    if (!table) return;
    ctd_emit(CTD_EV_SELECTION, table, (int64_t)info->iItem, 0);
}

static HWND ctd_table_window(ctd_handle table, ctd_status *problem) {
    HWND view = ctd_window(table);
    if (!view) { *problem = CTD_ERR_STALE; return NULL; }
    if (ctd_slot_kind(table) != CTD_W_TABLE) { *problem = CTD_ERR_KIND; return NULL; }
    *problem = CTD_OK;
    return view;
}

// -------------------------------------------------------------- entry points

ctd_status ctd_set_table_source(ctd_table_fn source, void *context) {
    g_table_source = source;
    g_table_context = context;
    return CTD_OK;
}

ctd_status ctd_set_table_edit_policy(ctd_table_editable_fn policy, void *context) {
    (void)policy; (void)context;
    return CTD_ERR_UNSUPPORTED;
}

ctd_status ctd_table_editing(ctd_handle table, int32_t on) {
    (void)table; (void)on;
    return CTD_ERR_UNSUPPORTED;
}

ctd_status ctd_table_edit_as_user(ctd_handle table, int32_t row, int32_t column,
                                  const char *utf8, int32_t len) {
    (void)table; (void)row; (void)column; (void)utf8; (void)len;
    return CTD_ERR_UNSUPPORTED;
}

ctd_status ctd_table_columns(ctd_handle table, int32_t count) {
    ctd_status problem;
    HWND view = ctd_table_window(table, &problem);
    if (!view) return problem;
    if (count < 0) return CTD_ERR_RANGE;
    HWND header = (HWND)SendMessageW(view, LVM_GETHEADER, 0, 0);
    int held = header ? (int)SendMessageW(header, HDM_GETITEMCOUNT, 0, 0) : 0;
    for (int i = held - 1; i >= 0; i--) {
        SendMessageW(view, LVM_DELETECOLUMN, (WPARAM)i, 0);
    }
    for (int32_t i = 0; i < count; i++) {
        LVCOLUMNW column;
        memset(&column, 0, sizeof column);
        column.mask = LVCF_TEXT | LVCF_WIDTH | LVCF_SUBITEM;
        column.pszText = L"";
        column.cx = 120;
        column.iSubItem = i;
        if (SendMessageW(view, LVM_INSERTCOLUMN, (WPARAM)i, (LPARAM)&column) < 0) {
            return CTD_ERR_PLATFORM;
        }
    }
    return CTD_OK;
}

ctd_status ctd_table_column_title(ctd_handle table, int32_t column,
                                  const char *utf8, int32_t len) {
    if (ctd_has_nul(utf8, len)) return CTD_ERR_RANGE;
    ctd_status problem;
    HWND view = ctd_table_window(table, &problem);
    if (!view) return problem;
    HWND header = (HWND)SendMessageW(view, LVM_GETHEADER, 0, 0);
    int held = header ? (int)SendMessageW(header, HDM_GETITEMCOUNT, 0, 0) : 0;
    if (column < 0 || column >= held) return CTD_ERR_RANGE;
    WCHAR *text = ctd_wide(utf8, len);
    if (!text) return CTD_ERR_PLATFORM;
    LVCOLUMNW item;
    memset(&item, 0, sizeof item);
    item.mask = LVCF_TEXT;
    item.pszText = text;
    LRESULT done = SendMessageW(view, LVM_SETCOLUMN, (WPARAM)column, (LPARAM)&item);
    free(text);
    return done ? CTD_OK : CTD_ERR_PLATFORM;
}

ctd_status ctd_table_column_width(ctd_handle table, int32_t column, double points) {
    ctd_status problem;
    HWND view = ctd_table_window(table, &problem);
    if (!view) return problem;
    HWND header = (HWND)SendMessageW(view, LVM_GETHEADER, 0, 0);
    int held = header ? (int)SendMessageW(header, HDM_GETITEMCOUNT, 0, 0) : 0;
    if (column < 0 || column >= held) return CTD_ERR_RANGE;
    if (points <= 0.0) return CTD_ERR_RANGE;
    SendMessageW(view, LVM_SETCOLUMNWIDTH, (WPARAM)column, (LPARAM)(int)(points + 0.5));
    return CTD_OK;
}

ctd_status ctd_table_rows(ctd_handle table, int32_t count) {
    ctd_status problem;
    HWND view = ctd_table_window(table, &problem);
    if (!view) return problem;
    if (count < 0) return CTD_ERR_RANGE;
    SendMessageW(view, LVM_SETITEMCOUNT, (WPARAM)count, LVSICF_NOSCROLL);
    return CTD_OK;
}

ctd_status ctd_table_reload(ctd_handle table) {
    ctd_status problem;
    HWND view = ctd_table_window(table, &problem);
    if (!view) return problem;
    // Everything an owner-data list holds is asked for again when it is
    // invalidated; there is nothing else to throw away.
    InvalidateRect(view, NULL, TRUE);
    return CTD_OK;
}

int32_t ctd_table_cell(ctd_handle table, int32_t row, int32_t column,
                       char *out, int32_t cap) {
    ctd_status problem;
    HWND view = ctd_table_window(table, &problem);
    if (!view) return problem;
    int rows = (int)SendMessageW(view, LVM_GETITEMCOUNT, 0, 0);
    if (row < 0 || row >= rows) return CTD_ERR_RANGE;
    HWND header = (HWND)SendMessageW(view, LVM_GETHEADER, 0, 0);
    int held = header ? (int)SendMessageW(header, HDM_GETITEMCOUNT, 0, 0) : 0;
    if (column < 0 || column >= held) return CTD_ERR_RANGE;
    // LVM_GETITEMTEXT on an owner-data list sends LVN_GETDISPINFO, so this
    // goes out through the source and back exactly the way a paint does.
    WCHAR room[512];
    room[0] = L'\0';
    LVITEMW item;
    memset(&item, 0, sizeof item);
    item.mask = LVIF_TEXT;
    item.iItem = row;
    item.iSubItem = column;
    item.pszText = room;
    item.cchTextMax = (int)(sizeof room / sizeof room[0]);
    SendMessageW(view, LVM_GETITEMTEXT, (WPARAM)row, (LPARAM)&item);
    return ctd_copy_wide_out(room, out, cap);
}

ctd_status ctd_table_selected(ctd_handle table, int32_t *out) {
    ctd_status problem;
    HWND view = ctd_table_window(table, &problem);
    if (!view) return problem;
    if (out) {
        *out = (int32_t)SendMessageW(view, LVM_GETNEXTITEM, (WPARAM)-1,
                                     (LPARAM)LVNI_SELECTED);
    }
    return CTD_OK;
}

ctd_status ctd_table_select(ctd_handle table, int32_t row) {
    ctd_status problem;
    HWND view = ctd_table_window(table, &problem);
    if (!view) return problem;
    LVITEMW item;
    memset(&item, 0, sizeof item);
    item.stateMask = LVIS_SELECTED | LVIS_FOCUSED;
    if (row < 0) {
        item.state = 0;
        g_writing++;
        SendMessageW(view, LVM_SETITEMSTATE, (WPARAM)-1, (LPARAM)&item);
        g_writing--;
        return CTD_OK;
    }
    int rows = (int)SendMessageW(view, LVM_GETITEMCOUNT, 0, 0);
    if (row >= rows) return CTD_ERR_RANGE;
    item.state = LVIS_SELECTED | LVIS_FOCUSED;
    g_writing++;
    SendMessageW(view, LVM_SETITEMSTATE, (WPARAM)row, (LPARAM)&item);
    g_writing--;
    return CTD_OK;
}
